pkg/controller/autoscaling/elasticsearch/controller.go

```diff
@@ -102,70 +159,219 @@ func (r *ReconcileElasticsearch) Reconcile(ctx context.Context, request reconcil
+	v, err := version.Parse(es.Spec.Version)
 	if err != nil {
-		return reconcile.Result{}, tracing.CaptureError(ctx, err)
-	}
-
-	if len(autoscalingSpecification.AutoscalingPolicySpecs) == 0 && len(autoscalingStatus.AutoscalingPolicyStatuses) == 0 {
-		// This cluster is not managed by the autoscaler
-		return reconcile.Result{}, nil
+		return reconcile.Result{}, err
 	}
-
-	// Get autoscaling policies and the associated node sets.
-	autoscaledNodeSets, nodeSetErr := autoscalingSpecification.GetAutoscaledNodeSets()
+	autoscaledNodeSets, nodeSetErr := es.GetAutoscaledNodeSets(v, esa.Spec.AutoscalingPolicySpecs)
 	if nodeSetErr != nil {
 		return reconcile.Result{}, tracing.CaptureError(ctx, nodeSetErr)
 	}
-	log.V(1).Info("Autoscaling policies and node sets", "policies", autoscaledNodeSets.Names())
+	log.V(1).Info(
+		"Autoscaling policies and node sets",
+		"policies", autoscaledNodeSets.Names(),
+		"namespace", request.Namespace,
+		"esa_name", request.Name,
+	)
 
 	// Import existing resources in the current Status if the cluster is managed by some autoscaling policies but
 	// the status annotation does not exist.
-	if err := autoscalingStatus.ImportExistingResources(log, r.Client, autoscalingSpecification, autoscaledNodeSets); err != nil {
+	if err := status.ImportExistingResources(
+		log,
+		r.Client,
+		esa.Spec.AutoscalingPolicySpecs,
+		es,
+		autoscaledNodeSets,
+		&esa.Status,
+	); err != nil {
+		_, _ = r.reportAsUnhealthy(ctx, log, esa, fmt.Sprintf("error while importing resources from the status subresource: %s", err.Error()))
+		// Status is updated on a best effort basis, we don't really care if the error above is not reported.
 		return reconcile.Result{}, tracing.CaptureError(ctx, err)
 	}
 
+	statusBuilder := newStatusBuilder(log, esa.Spec.AutoscalingPolicySpecs)
+	results := &reconciler.Results{}
+
 	// Call the main function
-	current, err := r.reconcileInternal(ctx, autoscalingStatus, autoscaledNodeSets, autoscalingSpecification, es)
+	reconciledEs, reconcileInternalErr := r.reconcileInternal(ctx, es, statusBuilder, autoscaledNodeSets, &esa)
+	if reconcileInternalErr != nil {
+		// we do not return immediately as not all errors prevent to compute a reconciled Elasticsearch resource.
+		results.WithError(reconcileInternalErr)
+	}
+
+	// Update the new status
+	newStatus := statusBuilder.Build()
+	esa.Status.ObservedGeneration = ptr.Int64(esa.Generation)
+	esa.Status.Conditions = esa.Status.Conditions.MergeWith(newStatus.Conditions...)
+	esa.Status.AutoscalingPolicyStatuses = newStatus.AutoscalingPolicyStatuses
+	updateStatus, err := r.updateStatus(ctx, log, esa)
 	if err != nil {
-		return reconcile.Result{}, tracing.CaptureError(ctx, err)
+		return reconcile.Result{}, err
 	}
+	results.WithResult(updateStatus)
+
+	if reconciledEs == nil {
+		// no reconciled Elasticsearch has been returned, likely to be the case if a fatal error prevented
+		// to calculate the new compute and storage resources.
```

Comment:
Might want to re-word this comment.

Reply:
I tried to improve this comment, not sure if it's enough though... let me know if you have a suggestion 🙇